### PR TITLE
Feature/csi 1552 cotroller fixs ds8k mediator

### PR DESCRIPTION
**Pull request by ArbelNathan**
separate ticket for finishing the review: CSI-1594

### CODE REVIEW

File: controller/array_action/array_mediator_ds8k.py [hunk 1]
```diff
@@ -29,6 +29,10 @@
 ERROR_CODE_VOLUME_NOT_FOUND_OR_ALREADY_PART_OF_CS_RELATIONSHIP = '00000013'
 MAX_VOLUME_LENGTH = 16
 
+FC_PERSISTENT_OPTION = "persistent"
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:32]: `FC` might be confused with Fibre Channel, rename to `FLASHCOPY_` or something

Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:32]: see [my previous comments](https://github.com/IBM/ibm-block-csi-driver/pull/173#discussion_r438202305) about those...

**Comment by ArbelNathan** [comment]
follow up - CSI-1594

File: controller/array_action/array_mediator_ds8k.py [hunk 2]
```diff
@@ -50,6 +54,19 @@ def get_volume_id_from_scsi_identifier(scsi_id):
     return scsi_id[-4:]
 
 
+def get_source_volume_id_if_exists(api_volume):
+    fc_source = [flashcopy.sourcevolume for flashcopy in api_volume.flashcopy if
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:58]: `fc_` might be confused with fibre channel. rename to `flashcopy_` or something (wherever applicable)

Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:58]: list var names should be plural

Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:58]: I would put the `if` in the same line with its condition

File: controller/controller_server/csi_controller_server.py [hunk 3]
```diff
@@ -418,7 +419,10 @@ def CreateSnapshot(self, request, context):
 
                 volume_name = array_mediator.get_volume_name(vol_id)
                 logger.info("Snapshot name : {}. Volume name : {}".format(snapshot_name, volume_name))
-                snapshot = array_mediator.get_snapshot(snapshot_name)
+                snapshot = array_mediator.get_snapshot(
+                    snapshot_name,
+                    volume_context=request.parameters
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:424]: I think we should avoid passing `volume_context` altogether, and pass `pool` instead (including existing code, e.g. `create_volume`)

File: controller/array_action/array_mediator_interface.py [hunk 4]
```diff
@@ -66,6 +66,7 @@ def copy_to_existing_volume_from_snapshot(self, name, src_snap_name, src_snap_ca
             src_snap_capacity_in_bytes   : capacity of snapshot to create from
             min_vol_size_in_bytes        : if snapshot capacity is lower than this value vol will
                                            be increased to this value
+            pool: pool of the volume and snapshot to find them more efficiently.
```
Comment (oriyarde) [inline controller/array_action/array_mediator_interface.py:69]: consider renaming `pool` to `pool_id`/`pool_name` (whichever is more accurate, wherever applicable)

File: controller/array_action/array_mediator_ds8k.py [hunk 5]
```diff
@@ -643,3 +676,11 @@ def _generate_snapshot_response(self, api_snapshot, source_volume_name):
     def validate_flashcopy(self, flashcopy_id):
         api_flashcopy = self._get_flashcopy(flashcopy_id)
         return api_flashcopy.state == 'valid'
+
+    def _check_snapshot_use_status(self, snapshot_id, flashcopy_list):
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:680]: consider renaming to something like `assert_snapshot_not_in_use`

File: controller/array_action/array_mediator_ds8k.py [hunk 6]
```diff
@@ -643,3 +676,11 @@ def _generate_snapshot_response(self, api_snapshot, source_volume_name):
...
+        for flashcopy in flashcopy_list:
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:681]: consider renaming `flashcopy_list` to `flashcopies`

File: controller/array_action/array_mediator_ds8k.py [hunk 7]
```diff
@@ -643,3 +676,11 @@ def _generate_snapshot_response(self, api_snapshot, source_volume_name):
...
+            logger.info("Deleting flashcopy: {}".format(flashcopy))
+            if flashcopy.sourcevolume == snapshot_id:
+                flashcopy_rel = self._get_flashcopy(flashcopy.id)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:684]: what is the difference between `flashcopy` and `flashcopy_rel` here?

File: controller/controller_server/csi_controller_server.py [hunk 8]
```diff
@@ -390,6 +390,7 @@ def ListVolumes(self, request, context):
 
     def CreateSnapshot(self, request, context):
         set_current_thread_name(request.name)
+        logger.info("Create snapshot with the request: {}".format(request))
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:393]: this logs our secret

File: controller/array_action/array_mediator_ds8k.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py]: solved in https://github.com/IBM/ibm-block-csi-driver/pull/201

File: controller/array_action/array_mediator_ds8k.py (hunk 2 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py]: `fcrel`+`fc_rel`+`flashcopy/ies` are better - so it's done

File: controller/array_action/array_mediator_ds8k.py (hunk 7 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py:684]: answered in: https://github.com/IBM/ibm-block-csi-driver/pull/200#discussion_r451531608

**Pull request merged**